pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -591,41 +585,38 @@ func (vs *VSphere) DisksAreAttached(volPaths []string, nodeName k8stypes.NodeNam
 		if err != nil {
 			return nil, err
 		}
-		vm, err := vs.getVMByName(ctx, nodeName)
+		dc, err := vclib.GetDatacenter(ctx, vs.conn, vs.cfg.Global.Datacenter)
 		if err != nil {
-			if vclib.IsNotFound(err) {
-				glog.Warningf("Node %q does not exist, vsphere CP will assume all disks %v are not attached to it.", nodeName, volPaths)
-				// make all the disks as detached and return false without error.
-				attached := make(map[string]bool)
-				for _, volPath := range volPaths {
-					attached[volPath] = false
-				}
-				return attached, nil
-			}
-			glog.Errorf("Failed to get VM object for node: %q. err: +%v", vSphereInstance, err)
 			return nil, err
 		}
 
-		for _, volPath := range volPaths {
-			result, err := vm.IsDiskAttached(ctx, volPath)
-			if err == nil {
-				if result {
-					attached[volPath] = true
-				} else {
-					attached[volPath] = false
+		vmVolumes := make(map[string][]string)
+		for nodeName, volPaths := range nodeVolumes {
+			for i, volPath := range volPaths {
+				canonicalVolumePath, err := getcanonicalVolumePath(ctx, dc, volPath)
+				if err != nil {
+					glog.Errorf("Failed to get canonical vsphere volume path for volume: %s. err: %+v", volPath, err)
+					return nil, err
+				}
+				if len(canonicalVolumePath) > 0 && filepath.Ext(canonicalVolumePath) != ".vmdk" {
```

Comment:
why do you need this check here?

Reply:
In case of static provisioning, the users can also have the volumePath without "vmdk" extension in the volume spec. For example, "[vsanDatastore] kubevols/redis-slave". This path will be used by kubernetes for "Is volume attach on node" check.

In this case, we need to check if the volume path doesn't have an extension, add an extension and compare with backing.FilePath on VM for correctness.

Reply:
I see. Please add this in comment. So that in future no one will remove this block by mistake.

Reply:
I have added comments.